src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LocalProxyVarsGen.java

```diff
@@ -0,0 +1,271 @@
+            super.visitClassDef(tree);
+        } finally {
+            currentClass = prevCurrentClass;
+            strictInstanceFields = prevStrictInstanceFields;
+        }
+    }
+
+    public void visitMethodDef(JCMethodDecl tree) {
+        if (strictFieldsReadInPrologue.get(tree) != null) {
+            Set<Symbol> fieldSet = strictFieldsReadInPrologue.get(tree);
+            if (!strictInstanceFields.isEmpty()) {
+                // ignore telescopic and generated constructors
+                if (TreeInfo.isConstructor(tree) &&
+                        TreeInfo.hasConstructorCall(tree, names._super) &&
+                        (tree.sym.flags_field & Flags.GENERATEDCONSTR) == 0) {
+                    // now we need to analyze the constructor's body, it could be that it is empty or that
+                    // no assignment to strict fields is done
+                    java.util.List<JCVariableDecl> strictFieldsRead = new ArrayList<>();
+                    for (JCVariableDecl sfield : strictInstanceFields) {
+                        if (fieldSet.contains(sfield.sym)) {
+                            strictFieldsRead.add(sfield);
+                        }
+                    }
+                    if (!strictFieldsRead.isEmpty()) {
+                        addLocalProxiesFor(tree, strictFieldsRead);
+                    }
+                }
+            }
+        }
+        super.visitMethodDef(tree);
+    }
+
+    void addLocalProxiesFor(JCMethodDecl constructor, java.util.List<JCVariableDecl> multiAssignedStrictFields) {
+        ListBuffer<JCStatement> localDeclarations = new ListBuffer<>();
+        Map<Symbol, Symbol> fieldToLocalMap = new LinkedHashMap<>();
+
+        for (JCVariableDecl fieldDecl : multiAssignedStrictFields) {
+            long flags = SYNTHETIC;
+            VarSymbol proxy = new VarSymbol(flags, newLocalName(fieldDecl.name.toString()), fieldDecl.sym.erasure(types), constructor.sym);
+            fieldToLocalMap.put(fieldDecl.sym, proxy);
+            JCVariableDecl localDecl = make.at(constructor.pos).VarDef(proxy, fieldDecl.init);
+            localDecl.vartype = fieldDecl.vartype;
+            localDeclarations = localDeclarations.append(localDecl);
+        }
+
+        FieldRewriter fr = new FieldRewriter(constructor, fieldToLocalMap);
+        ListBuffer<JCStatement> newBody = new ListBuffer<>();
+        for (JCStatement st : constructor.body.stats) {
+            newBody = newBody.append(fr.translate(st));
+        }
+        localDeclarations.addAll(newBody);
+        ListBuffer<JCStatement> assigmentsBeforeSuper = new ListBuffer<>();
+        for (Symbol vsym : fieldToLocalMap.keySet()) {
+            Symbol local = fieldToLocalMap.get(vsym);
+            assigmentsBeforeSuper.append(make.at(constructor.pos()).Assignment(vsym, make.at(constructor.pos()).Ident(local)));
+        }
+        constructor.body.stats = localDeclarations.toList();
+        if (!assigmentsBeforeSuper.isEmpty()) {
```

Comment:
Can this ever be empty (given we end up here only when `Resolve` found something) ?